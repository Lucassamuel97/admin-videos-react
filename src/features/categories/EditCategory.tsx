import { Box, Paper, Typography } from "@mui/material";
import { useState } from "react";
import { useParams } from "react-router-dom";
import { useAppSelector } from "../../app/hooks";
import { selectCategoryById } from "./categorySlice";
import { CategoryForm } from "./components/CategoryForm";

export const CategoryEdit = () => {

  const id = useParams().id || "";
  const [isDisabled, setIsDisabled] = useState(false);
  const category = useAppSelector((state) => selectCategoryById(state, id));

  const handleChange = (e: any) => { };
  const handleToggle = (e: any) => { };

  return (
    <Box>
      <Paper>
        <Box p={2}>
          <Box mb={2}>
            <Typography variant="h4">Edit Category</Typography>
          </Box>
          <CategoryForm
            category={category}
            isDisabled={isDisabled}
            onSubmit={(e) => {
              e.preventDefault();
              // handle submit
            }}
            handleChange={handleChange}
            handleToggle={handleToggle}
          />
        </Box>
      </Paper>
    </Box>
  );
};