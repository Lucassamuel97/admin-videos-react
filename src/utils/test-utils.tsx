import { ReactElement, PropsWithChildren } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { SnackbarProvider } from "notistack";
// Removed PreloadedState as it is not exported from @reduxjs/toolkit

import { AppStore, RootState, makeStore } from "../app/store";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: ReactElement,
  {
    preloadedState,
    store = makeStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  const Wrapper = ({ children }: PropsWithChildren<{}>): JSX.Element => (
    <Provider store={store}>
      <BrowserRouter>
        <SnackbarProvider maxSnack={3} autoHideDuration={3000}>
          {children}
        </SnackbarProvider>
      </BrowserRouter>
    </Provider>
  );

  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}

export * from "@testing-library/react";
