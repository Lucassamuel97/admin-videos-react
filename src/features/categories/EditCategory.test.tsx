import { http, HttpResponse } from "msw";
import { setupServer } from "msw/node";
import { renderWithProviders, screen, waitFor, fireEvent } from "../../utils/test-utils";
import { CategoryEdit } from "./EditCategory";
import { baseUrl } from "../api/apiSlice";

const data = {
    id: "1",
    name: "Category 1",
    description: "Category test",
    is_active: true,
    deleted_at: null,
    created_at: "2022-09-27T17:10:33+0000",
    updated_at: "2022-09-27T17:10:33+0000",
};

export const handlers = [
    http.get(`${baseUrl}/categories/:id`, async ({ params }) => {
        return HttpResponse.json({data});
    }),
    http.put(`${baseUrl}/categories/${data.id}`, async () => {
        await new Promise(resolve => setTimeout(resolve, 200));
        return HttpResponse.json(
            {data},
            { status: 201 }
        );
    }
    ),
];

const server = setupServer(...handlers);

describe("EditCategory", () => {
    afterAll(() => server.close());
    beforeAll(() => server.listen());
    afterEach(() => server.resetHandlers());

    it("should render correctly", () => {
        const { asFragment } = renderWithProviders(<CategoryEdit />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("should handle submit", async () => {
        renderWithProviders(<CategoryEdit />);
        const name = await screen.findByTestId("name");
        const description = screen.getByTestId("description");
        const isActive = screen.getByTestId("is_active");
        const submit = screen.getByText("Save");

        await waitFor(() => {
            expect(name).toHaveValue("Category 1");
        });

        fireEvent.change(name, { target: { value: "test" } });
        fireEvent.change(description, { target: { value: "test desc" } });
        fireEvent.click(isActive);
        fireEvent.click(submit);

        await waitFor(() => {
            const text = screen.getByText("Category updated successfully");
            expect(text).toBeInTheDocument();
        });
    
    });

    it("should handle error", async () => {
        server.use(
            http.put(`${baseUrl}/categories/${data.id}`, async () => {
                return HttpResponse.json(
                    { error: "Algo deu errado!" },  // Corpo da resposta
                    { status: 500 }  // Status HTTP 500
                );
            })
        );

        renderWithProviders(<CategoryEdit />);
        const name = await screen.findByTestId("name");
        const description = screen.getByTestId("description");
        const isActive = screen.getByTestId("is_active");
        const submit = screen.getByText("Save");

        fireEvent.change(name, { target: { value: "test" } });
        fireEvent.change(description, { target: { value: "test desc" } });
        fireEvent.click(isActive);
        fireEvent.click(submit);

        await waitFor(() => {
            const text = screen.getByText("Category not updated");
            expect(text).toBeInTheDocument();
        });
    });
});