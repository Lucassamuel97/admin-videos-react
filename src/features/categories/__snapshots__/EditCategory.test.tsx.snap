// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditCategory > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1gtchvp-MuiPaper-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="MuiBox-root css-19midj6"
      >
        <div
          class="MuiBox-root css-1qm1lh"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-f84ov1-MuiTypography-root"
          >
            Edit Category
          </h4>
        </div>
      </div>
      <div
        class="MuiBox-root css-1qm1lh"
      >
        <form>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-1qw2a80-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth css-ytlejw-MuiFormControl-root"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-disabled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for="name"
                    id="name-label"
                  >
                    Name
                    <span
                      aria-hidden="true"
                      class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ljffdk-MuiFormLabel-asterisk"
                    >
                       *
                    </span>
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                      data-testid="name"
                      disabled=""
                      id="name"
                      name="name"
                      required=""
                      type="text"
                      value=""
                    />
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-81qg8w"
                      >
                        <span>
                          Name *
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth css-ytlejw-MuiFormControl-root"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-disabled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for="description"
                    id="description-label"
                  >
                    Description
                    <span
                      aria-hidden="true"
                      class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ljffdk-MuiFormLabel-asterisk"
                    >
                       *
                    </span>
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                      data-testid="description"
                      disabled=""
                      id="description"
                      name="description"
                      required=""
                      type="text"
                      value=""
                    />
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-81qg8w"
                      >
                        <span>
                          Description *
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiFormGroup-root css-1mac7pt-MuiFormGroup-root"
              >
                <label
                  class="MuiFormControlLabel-root Mui-disabled MuiFormControlLabel-labelPlacementEnd css-19opgx6-MuiFormControlLabel-root"
                >
                  <span
                    class="MuiSwitch-root MuiSwitch-sizeMedium css-rehya9-MuiSwitch-root"
                  >
                    <span
                      aria-disabled="true"
                      class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorSecondary Mui-disabled Mui-disabled PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorSecondary Mui-disabled Mui-disabled css-zp7cnf-MuiButtonBase-root-MuiSwitch-switchBase"
                      data-testid="is_active"
                      tabindex="-1"
                    >
                      <input
                        aria-label="controlled"
                        class="PrivateSwitchBase-input MuiSwitch-input css-j8yymo"
                        disabled=""
                        name="is_active"
                        type="checkbox"
                      />
                      <span
                        class="MuiSwitch-thumb css-17jyosd-MuiSwitch-thumb"
                      />
                    </span>
                    <span
                      class="MuiSwitch-track css-tj8d99-MuiSwitch-track"
                    />
                  </span>
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label Mui-disabled css-rizt0-MuiTypography-root"
                  >
                    Active
                  </span>
                </label>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiBox-root css-10egq61"
              >
                <a
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
                  href="/categories"
                  tabindex="0"
                >
                  Back
                </a>
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary css-1mns5ui-MuiButtonBase-root-MuiButton-root"
                  disabled=""
                  tabindex="-1"
                  type="submit"
                >
                  Save
                </button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
